src/main/java/explorableviz/transparenttext/paragraph/Paragraph.java

```diff
@@ -18,59 +14,78 @@ public Paragraph() {
     }
 
-    public void spliceExpression(String expression) {
-        List<TextFragment> paragraph = this;
-        ListIterator<TextFragment> iterator = paragraph.listIterator();
+    private Pair<Expression, Paragraph> buildParagraph(Paragraph template, int numComputedExpr, int index) {
+        Paragraph p = new Paragraph();
+        Expression toCompute = null;
+        int exprCount = 0;
+        StringBuilder literalBuilder = new StringBuilder();
+        int replaceStart = -1, replaceEnd = -1;
 
-        while (iterator.hasNext()) {
-            TextFragment textFragment = iterator.next();
-            if (textFragment instanceof Literal) {
-                splitLiteral(textFragment).ifPresentOrElse(expectedValue -> {
-                    iterator.remove();
-                    iterator.add(expectedValue.beforeTag());
-                    iterator.add(new Expression(expression, expectedValue.tag().getValue()));
-                    iterator.add(expectedValue.afterTag());
-                }, () -> {
-                    throw new RuntimeException("REPLACE tag not found");
-                });
+        for (int k = 0; k < template.size(); k++) {
+            var element = template.get(k);
+
+            if (element instanceof Literal literal) {
+                literalBuilder.append(" ").append(literal.getValue());
+            } else if (element instanceof Expression expression) {
+                if (exprCount < numComputedExpr) {
+                    Expression computedExpr = (Expression) this.get(k);
+                    p.add(new Literal(literalBuilder.toString(), null));
+                    literalBuilder.setLength(0);
+                    p.add(computedExpr != null ? computedExpr : expression);
+                } else if (exprCount == numComputedExpr + index) {
+                    replaceStart = literalBuilder.length();
+                    literalBuilder.append(" ").append(expression.getValue());
+                    replaceEnd = literalBuilder.length();
+                    toCompute = expression;
+                } else {
+                    literalBuilder.append(" ").append(expression.getValue());
+                }
+                exprCount++;
             }
         }
+
+        p.add(new Literal(literalBuilder.toString(), new Literal.SelectedRegion(replaceStart, replaceEnd)));
+        return new Pair<>(toCompute, p);
     }
 
-    public static Optional<LiteralParts> splitLiteral(TextFragment literal) {
-        Matcher valueReplaceMatcher = Pattern.compile("(.*)\\[REPLACE id=\".*?\" value=\"(.*?)\"](.*)").matcher(literal.getValue());
-        if (!valueReplaceMatcher.find()) {
-            return Optional.empty();
+    public Paragraph splice(Expression expression) {
+        Paragraph p = new Paragraph();
+        for (TextFragment t : this) {
+            if(t instanceof Literal l && l.getSelectedRegion() != null) {
+                //@todo special cases - if the literal.lenght =0 - skip
```

Comment:
These special cases need doing as a literal of length zero can't arise in a Fluid program and so this won't represent a real AST.